src/makefile.ts

```diff
@@ -0,0 +1,97 @@
+import { FileBase, FileBaseOptions, IResolver } from './file';
+import { Project } from './project';
+
+/**
+ * A Make rule.
+ */
+export interface Rule {
+  /**
+   * Files to be created or updated by this rule.
+   *
+   * If the rule is phony then instead this represents the command's name(s).
+   */
+  readonly targets: string[];
+
+  /**
+   * Files that are used as inputs to create a target.
+   */
+  readonly prerequisites?: string[];
+
+  /**
+   * Commands that are run (using prerequisites as inputs) to create a target.
+   */
+  readonly recipe?: string[];
+
+  /**
+   * Marks whether the target is phony.
+   *
+   * False by default.
+   */
+  readonly phony?: boolean;
+}
+
+/**
+ * Options for Makefiles.
+ */
+export interface MakefileOptions extends FileBaseOptions {
+  /**
+   * List of targets to build when Make is invoked without specifying any targets.
+   */
+  readonly all?: string[];
+
+  /**
+   * Rules to include in the Makefile.
+   */
+  readonly rules: Rule[];
+}
+
+/**
+ * Minimal Makefile.
+ */
+export class Makefile extends FileBase {
+  /**
+   * Targets to make by default.
+   */
+  public readonly all?: string[];
+
+  /**
+   * List of rule definitions.
+   */
+  public readonly rules: Rule[];
+
+
+  constructor(project: Project, filePath: string, options: MakefileOptions) {
+    super(project, filePath, options);
+
+    if (!options.rules) {
+      throw new Error('"rules" cannot be undefined');
+    }
+
+    this.all = options.all;
+    this.rules = options.rules;
```

Comment:
You can just implement `all` through `rules` I believe:

```suggestion
    const all = { phony: true, targets: ['all'], prerequisites: options.all };
    this.rules = [
       ...options.all ? all : [],
       ...options.rules,
    ];
```

Reply:
done

Reply:
Also it turns out
```
.PHONY: all
all:
```
is a valid Makefile, so I've rewritten it to always add the all rule